snapshot/btrfs/btrfs.go

```diff
@@ -6,63 +6,152 @@ import (
+func (b *Snapshotter) Stat(key string) (snapshot.Info, error) {
+	// resolve the snapshot out of the index.
+	target, err := os.Readlink(filepath.Join(b.root, "index", hash(key)))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return snapshot.Info{}, err
+		}
+
+		return snapshot.Info{}, errors.Errorf("snapshot %v not found", key)
+	}
+
+	return b.stat(target)
+}
+
+func (b *Snapshotter) stat(target string) (snapshot.Info, error) {
+	var (
+		parents    = filepath.Join(b.root, "parents")
+		names      = filepath.Join(b.root, "names")
+		namep      = filepath.Join(names, filepath.Base(target))
+		parentlink = filepath.Join(parents, filepath.Base(target))
+	)
+
+	// grab information about the subvolume
+	info, err := btrfs.SubvolInfo(target)
+	if err != nil {
+		return snapshot.Info{}, err
+	}
+
+	// read the name out of the names!
+	nameraw, err := ioutil.ReadFile(namep)
+	if err != nil {
+		return snapshot.Info{}, err
+	}
+
+	// resolve the parents path.
+	parentp, err := os.Readlink(parentlink)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return snapshot.Info{}, err
+		}
+
+		// no parent!
+	}
+
+	var parent string
+	if parentp != "" {
+		// okay, grab the basename of the parent and look up its name!
+		parentnamep := filepath.Join(names, filepath.Base(parentp))
+
+		p, err := ioutil.ReadFile(parentnamep)
+		if err != nil {
+			return snapshot.Info{}, err
+		}
+
+		parent = string(p)
+	}
+
+	kind := snapshot.KindCommitted
+	if strings.HasPrefix(target, filepath.Join(b.root, "active")) {
+		kind = snapshot.KindActive
+	}
+
+	return snapshot.Info{
+		Name:     string(nameraw),
+		Parent:   parent,
+		Readonly: info.Readonly,
+		Kind:     kind,
+	}, nil
+
 }
 
-func (b *Driver) Prepare(key, parent string) ([]containerd.Mount, error) {
+func (b *Snapshotter) Prepare(key, parent string) ([]containerd.Mount, error) {
 	return b.makeActive(key, parent, false)
 }
 
-func (b *Driver) View(key, parent string) ([]containerd.Mount, error) {
+func (b *Snapshotter) View(key, parent string) ([]containerd.Mount, error) {
 	return b.makeActive(key, parent, true)
 }
 
-func (b *Driver) makeActive(key, parent string, readonly bool) ([]containerd.Mount, error) {
+func (b *Snapshotter) makeActive(key, parent string, readonly bool) ([]containerd.Mount, error) {
 	var (
 		active     = filepath.Join(b.root, "active")
-		parents    = filepath.Join(b.root, "parents")
 		snapshots  = filepath.Join(b.root, "snapshots")
+		parents    = filepath.Join(b.root, "parents")
+		index      = filepath.Join(b.root, "index")
 		names      = filepath.Join(b.root, "names")
 		keyh       = hash(key)
 		parenth    = hash(parent)
-		dir        = filepath.Join(active, keyh)
+		target     = filepath.Join(active, keyh)
 		namep      = filepath.Join(names, keyh)
+		indexlink  = filepath.Join(index, keyh)
 		parentlink = filepath.Join(parents, keyh)
-		parentp    = filepath.Join(snapshots, parenth)
+		parentp    = filepath.Join(snapshots, parenth) // parent must be restricted to snaps
```

Comment:
(fairly minor and I'm betting you'll disagree, but...) I prefer things like `parentPath` and `keyHash` to `parentp` and `keyh` respectively.  When reading diffs to review code, having names instead of just abbreviations helps me to understand without having to read more context than is involved in the immediate code change.